CFE-4590, ENT-13239: Override default directory permissions

larsewi opened the pull request:
- **Added test for overriding default directory create mode**
- **Skip remaining string comparisons after successful match**
- **Added override_default_directory_create_mode to body agent control**
- **Replaced hardcoded directory create mode with DEFAULTMODE**

larsewi committed:
Added test for overriding default directory create mode

Ticket: CFE-4590, ENT-13239
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Skip remaining string comparisons after successful match

Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Added default_directory_create_mode to body agent control

Allow overriding the default 0700 permissions when `cf-agent` creates
parent directories during file promise repairs.

The new attribute `default_directory_create_mode` in body agent control
enables users to specify custom permissions (e.g., 0755) for
automatically created directories, avoiding the need for explicit perms
promises on each parent directory when deeper paths are required.

This addresses cases where files need broader access permissions but
their auto-created parent directories would otherwise default to 0700,
making the files inaccessible despite having correct permissions.

Example usage:
```
body agent control {
    default_directory_create_mode => "0755";
}
```

Ticket: CFE-4590, ENT-13239
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Replaced hardcoded directory create mode with DEFAULTMODE

Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=12934)](https://ci.cfengine.com//job/pr-pipeline/12934/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/12934/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-12934/

craigcomstock reviewed (approved):
looks good.

nickanderson commented on cf-agent/cf-agent.c:
```diff
@@ -1337,6 +1337,7 @@ static void KeepControlPromises(EvalContext *ctx, const Policy *policy, GenericA
             {
                 Log(LOG_LEVEL_VERBOSE, "SET select_end_match_eof %s", (char *) value);
                 EvalContextSetSelectEndMatchEof(ctx, BooleanFromString(value));
+                continue;
```
Feels like i am just showing what i dont know about the code

but, @larsewi your description there made me wonder if this would be affected:

```
body common control
{
   default_directory_create_mode => "000";
   default_directory_create_mode => "770";
}
```
It's a silly example, but in some policies the same attribtue is set differently in different contexts and multiple contexts might apply. 

I just wanted to be sure that not continuing to check for the attribute will not affect this (in the above simple example, expect the last one to win).

larsewi replied:
I don't think these `break`s will affect multiple configurations of the default directory create mode. However, I added a line in the policy to test this, just in case. I will merge once the acceptance test workflow from GitHub Actions have passed.

larsewi committed:
Make sure the last configuration wins

When default create mode is configured multiple times, then make sure
the last configuration always wins.

Ticket: CFE-4590, ENT-13239
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>
